File: lib/poker_node/hand.rb
require 'set'

module PokerNode
  class Hand
    attr_reader :cards

    attr_reader :suit
    attr_reader :kind
    attr_reader :flush_suit
    attr_reader :straight_kind
    
    def initialize(cards)
      @cards = cards.is_a?(String) ? Card.parse!(cards) : cards
    end

    def kind
      @kind ||= @cards.group_by(&:kind)
    end

    def kinds
      @kinds ||= @cards.map(&:kind)
    end

    def kinds_set
      Set.new(kinds)
    end

    def straight_kind
      Kind.straights.each { |kinds|
        @straight_kind = kinds.last if Set.new(kinds).subset?(kinds_set)
      }
      @straight_kind
    end

    def suit
      @suit ||= @cards.group_by(&:suit)
    end
    
    def suits
      @suits ||= @cards.map(&:suit)
    end

    def flush_suit
      @flush_suit ||= suits.select { |suit| self.suit[suit].size >= 5 }.first
    end

    def cards_by_count(n)
      kind.values.select { |cards| cards.size == n }
    end

    def high_card?
      true
    end

    def straight_flush?
      straight? && flush?
    end

    def royal_flush?
      straight_flush? && straight_kind == 'A'
    end

    def flush?
      flush_suit.present?
    end

    def straight?
      straight_kind.present?
    end
    
    def wheel_straight?
      straight? && straight_kind == '5'
    end

    def quads
      cards_by_count(4)
    end

    def quad
      quads.first
    end

    def four_of_kind?
      quad.present?
    end

    def full_house?
      set? && pair?
    end

    def sets
      cards_by_count(3)
    end

    def set
      sets.first
    end

    def three_of_kind?
      set.present?
    end

    alias :set? three_of_kind?

    def pairs
      @pairs ||= cards_by_count(2)
    end

    def pairs_sorted
      pairs.sort_by(&:max)
    end

    def pair
      pairs_sorted.first
    end

    def two_pair? # at least 2 pairs
      pairs.size >= 2
    end

    def one_pair? # at least 1 pair
      !pairs.empty?
    end

    def query?(hand)
      send("#{hand}?")
    end
    
    alias :pair? one_pair?
  end
  
  class HighCard
    include Comparable
    
    attr_reader :hand
    attr_reader :high_cards
    attr_reader :kickers

    def initialize(hand, hole)
      @hand = hand
      @hole = hole
    end

    def hole_cards
      @hole.cards
    end

    def high_cards
      @hole.cards
    end

    def kickers
      @hand.cards - high_cards
    end

    def name
      self.class.to_s.demodulize.underscore
    end

    def explain
      "High card #{@hole.cards.max}"
    end

    def inspect
      "<Rank:#{explain}\n\thole_cards=#{hole_cards}\n\thigh_cards=#{high_cards}\n\tkickers=#{kickers}>"
    end

    def index
      PokerNode::Poker::HIGH.index(name.to_sym)
    end

    def <=>(other)
      index_compare = self.index <=> other.index
      return index_compare unless index_compare == 0
      self.high_cards.each_with_index { |card, i|
        high_card_compare = card <=> other.high_cards[i]
        return high_card_compare unless high_card_compare == 0
      }
      self.kickers.each_with_index { |card, i|
        kicker_compare = card <=> other.kickers[i]
        return kicker_compare unless kicker_compare == 0
      }
      return 0
    end
  end

  class Flush < HighCard
    def high_cards
      @hand.suit[@hand.flush_suit]
    end

    def explain
      "Flush of #{@hand.flush_suit}"
    end
  end

  class Straight < HighCard
    def high_cards
      @hand.kind[@hand.straight_kind]
    end

    def explain
      "Straight to #{@hand.straight_kind}"
    end
  end

  class StraightFlush < Straight
    def explain
      "Straight flush of #{@hand.flush_suit} to #{@hand.straight_kind}"
    end
  end

  class FourOfKind < HighCard
    def high_cards
      @hand.quad
    end

    def explain
      "Four of #{@hand.quad.max.kind}'s"
    end
  end

  class ThreeOfKind < HighCard
    def high_cards
      @hand.set
    end

    def explain
      "Three of #{high_cards.max.kind}'s"
    end
  end

  class FullHouse < ThreeOfKind
    def explain
      "#{@hand.set.max.kind}'s full of #{@hand.pair.max.kind}'s"
    end
  end

  class OnePair < HighCard
    def high_cards
      @hand.pair
    end

    def explain
      "Pair of #{@hand.pair.max.kind}'s"
    end
  end

  class TwoPair < OnePair
    def high_cards
      @hand.pair
    end

    def explain
      "Pairs of #{@hand.pairs_sorted.first.max.kind}'s and #{@hand.pairs_sorted.second.max.kind}'s"
    end
  end
end

File: js/pokernode/hand.js
Rank = function(data) {
    this.kind = data.kind
    this.suit = data.suit
    this.cards = data.cards

    this.toString = function() {
        return "[kind=" + this.kind + " suit=" + Suit.draw(this.suit) + " cards=" + this.cards.toString() + "]"
    }
}

Hand = function(cards) {
    this.suits = {}
    for (var i = 0; i < cards.length; i++) {
        var card = cards[i]
        if (typeof(this.suits[card.suit]) == 'undefined') {
            this.suits[card.suit] = []
        }
        this.suits[card.suit].push(card)
    }
    
    this.kinds = {}
    for (i = 0; i < cards.length; i++) {
        card = cards[i]
        if (typeof(this.kinds[card.kind]) == 'undefined') {
            this.kinds[card.kind] = []
        }
        this.kinds[card.kind].push(card)
    }
    this.cards = cards

    this.isStraightFlush = function() {
        if (this.isStraight()) {
            var value = this.value
            var cards = value[1].cards
            var suit = null
            for (var i = 0; i < cards.length; i++) {
                var card = cards[i]
                if (suit == null)
                    suit = card.suit
                else if (suit != card.suit)
                    return false
            }
            this.value[0] = 'straight_flush'
            this.value[1].suit = suit
            return true
        }
        return false
    }

    this.isStraight = function() {
        for (var i = 8; i >= -1; i--) {
            var slice = null
            if (i == -1) {
                slice = ['A'].concat(Kind.all.slice(0, 4))
            } else {
                slice = Kind.all.slice(i, i + 5)
            }
            var result = true
            var cards = []
            for (var j = 0; j < 5; j++) {
                var kind = slice[j]
                var present = false
                for (var k = 0; k < this.cards.length; k++) {
                    var card = this.cards[k]
                    if (card.kind == kind) {
                        cards.push(card)
                        present = true
                    }
                }
                if (!present) {
                    result = false
                    break
                }
            }
            if (result) {
                kind = Kind.all[i + 5]
                this.value = ['straight', new Rank({kind: kind, cards: cards})]
                return true
            }
        }
        return false
    }

    this.isFlush = function() {
        for (var i = 0; i < Suit.all.length; i++) {
            var suit = Suit.all[i]
            var cards = this.suits[suit]
            if (typeof(cards) == 'undefined')
                continue
            if (cards.length >= 5) {
                this.value = ['flush', new Rank({suit: suit, cards: cards})]
                return true
            }
        }
        return false
    }

    this.isQuad = function() {
        for (var i = 0; i < Kind.all.length; i++) {
            var kind = Kind.all[i]
            var cards = this.kinds[kind]
            if (typeof(cards) == 'undefined')
                continue
            if (cards.length == 4) {
                this.value = ['quad', new Rank({kind: kind, cards: cards})]
                return true
            }
        }
        return false
    }

    this.isFullHouse = function() {
        if (this.isSet()) {
            var value = this.value
            if (this.isPair()) {
                this.value = ['full_house', [value[1], this.value[1]]]
                return true
            }
        }
        return false
    }

    this.isSet = function() {
        for (var i = 0; i < Kind.all.length; i++) {
            var kind = Kind.all[i]
            var cards = this.kinds[kind]
            if (typeof(cards) == 'undefined')
                continue
            if (cards.length == 3) {
                this.value = ['set', new Rank({kind: kind, cards: cards})]
                return true
            }
        }
        return false
    }

    this.isDoper = function() {
        var total = 0
        var value = []
        for (var i = 0; i < Kind.all.length; i++) {
            var kind = Kind.all[i]
            var cards = this.kinds[kind]
            if (typeof(cards) == 'undefined')
                continue
            if (cards.length == 2) {
                value.push([kind, cards])
                total++
            }
        }
        if (total == 2) {
            this.value = ['doper', []]
            for (i = 0; i < value.length; i++) {
                this.value[1].push(new Rank({kind: value[i][0], cards: value[i][1]}))
            }
            return true
        }
        return false
    }

    this.isPair = function() {
        for (var i = 0; i < Kind.all.length; i++) {
            var kind = Kind.all[i]
            var cards = this.kinds[kind]
            if (typeof(cards) == 'undefined')
                continue
            if (cards.length == 2) {
                this.value = ['pair', new Rank({kind: kind, cards: cards})]
                return true
            }
        }
        return false
    }

    this.isHighCard = function() {
        this.value = ['high_card', new Rank({cards:this.cards})]
        return true
    }
}

Hand.detect = function(hand) {
    while(1) {
        if (hand.isStraightFlush()) break
        if (hand.isQuad()) break
        if (hand.isFullHouse()) break
        if (hand.isFlush()) break
        if (hand.isStraight()) break
        if (hand.isSet()) break
        if (hand.isDoper()) break
        if (hand.isPair()) break
        if (hand.isHighCard()) break
    }
    return hand.value[0]
}
